Fix for OpenVINO 2024.0. OpenVINO 2022.1 as minimal supported version.

olpipi opened the pull request:
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

dkurt commented on modules/dnn/src/ie_ngraph.hpp:
```diff
@@ -17,7 +17,11 @@
 #pragma warning(disable : 4245)
 #pragma warning(disable : 4268)
 #endif
+#if INF_ENGINE_VER_MAJOR_GE(INF_ENGINE_RELEASE_2024_0)
```
Should be valid even for earlier release: https://github.com/opencv/opencv/blob/0fabe2b9e9fac96957b553b6a83483bddfe129bb/modules/dnn/src/op_inf_engine.hpp#L49.

Not sure about `ov::` namespace

dkurt commented:
Thanks, @olpipi! Our CI uses OpenVINO 2023.0.2.

Can you determine on which minimal version `ov::` namespace and required headers are available? To understand which would be a minimal required version to support.

dkurt commented on modules/dnn/src/ie_ngraph.cpp:
```diff
@@ -401,9 +406,9 @@ InfEngineNgraphNode::InfEngineNgraphNode(const std::vector<Ptr<BackendNode> >& n
+    ov::OutputVector inp_nodes;
 #else
-    ngraph::NodeVector inp_nodes;
+    ov::NodeVector inp_nodes;
```
Is `ov::` available for `< 2020.3`?

ilya-lavrenov replied:
do you need to support 4 years old OpenVINO?
Can we say only 2022.1 and newer are supported? It will significantly simplify conditions in code where OpenVINO is used and also will allow to drop legacy cmake code and rely only on `find_package(OpenVINO)`

dkurt replied:
Agree. For now CI works on 2021.4 so we can keep it to test it against OpenVINO API 2.0 (2022.x). I just mean that this lane changed under `#ifdef` and we don't test it. So specifically for this line, I recommend revert. Let's drop OpenVINO support for 2020.x in a separate PR.

ilya-lavrenov replied:
>  2021.4

API 2.0 since 2022.1
If we want to drop IE:: and API 1.0, we need to test since 2022.1

ilya-lavrenov replied:
@dkurt are you OK to drop releases older 2022.1 ? to keep API 2.0 only

dkurt replied:
I don't mind. Created an issue here: https://github.com/opencv/opencv/issues/25207.

This way https://pullrequest.opencv.org/buildbot/builders/4_x_openvino-opencl-skl-lin64 should be moved to 2022.1 or newer (/cc @asmorkalov, @opencv-alalek)

I only recommend to do this in a separate PR

ilya-lavrenov replied:
> I only recommend to do this in a separate PR

But in this case we need to do double job:
- Current PR supports both Pre and Post API 1.0 changes
- Next PR removes Pre API 1.0 code path.

I suppose we need to do it in one shot and just drop API 1.0 here

dkurt replied:
Sure, @olpipi, will you handle OpenVINO API 1.0 removal? Please keep 2022.1 is a minimal supported version.

olpipi commented:
> Thanks, @olpipi! Our CI uses OpenVINO 2023.0.2.
> 
> Can you determine on which minimal version `ov::` namespace and required headers are available? To understand which would be a minimal required version to support.

Hi
No problem. I'll check it

dkurt commented:
Please note that as a part of regular CI we test 2021.4.2 release (OpenVINO API 1.0): https://pullrequest.opencv.org/buildbot/builders/4_x_openvino-opencl-skl-lin64/builds/100055

If there is no `ov` namespace in 2021.4.2, I still recommend to replace `ngraph::` to `ov::` however with a workaround for old release:
```cpp
namespace ov = ngraph;
```

dkurt commented on modules/dnn/src/layers/elementwise_layers.cpp:
```diff
@@ -64,6 +64,19 @@
+#include <openvino/op/relu.hpp>
+#include <openvino/op/sigmoid.hpp>
+#include <openvino/op/sqrt.hpp>
+#include <openvino/op/tanh.hpp>
```
```cpp
#include <openvino/op/ops.hpp>
```

dkurt replied:
And I think better to include it to `op_inf_engine.hpp`

olpipi replied:
Include of separated ops speeds up build. But I cannot use this approach with 2021 OpenVINO release :(

olpipi commented:
@dkurt OpenVINO 2021.4.2  has neither ov namespace nor "openvino/..." include pathes. So I'll fix it as you suggested

dkurt commented on modules/dnn/src/ie_ngraph.hpp:
```diff
@@ -17,7 +17,16 @@
+#include <openvino/op/ops.hpp>
+#else
 #include <ngraph/ngraph.hpp>
+namespace ov = ngraph;
```
Conflicts with existing namespace

ilya-lavrenov commented on modules/dnn/src/layers/detection_output_layer.cpp:
```diff
@@ -55,11 +55,7 @@
 
 #ifdef HAVE_DNN_NGRAPH
```
HAVE_DNN_NGRAPH => HAVE_DNN_OPENVINO ?

dkurt replied:
Will cost a lot of changes as well as DNN_BACKEND_INFERENCE_ENGINE -> DNN_BACKEND_OPENVINO.

ilya-lavrenov commented on modules/dnn/src/net_openvino.cpp:
```diff
@@ -9,6 +9,12 @@
 
+#include "op_inf_engine.hpp"
+
+#ifdef HAVE_INF_ENGINE
```
```suggestion
#ifdef HAVE_OPENVINO
```

ilya-lavrenov commented on modules/dnn/src/net_openvino.cpp:
```diff
@@ -864,15 +804,8 @@ Net openvino_readNetwork(
     InferenceEngine::CNNNetwork ieNet;
```
it's removed in API 2.0

dkurt replied:
This is a workaround to wrap `ov::Model` from API 2.0: https://github.com/opencv/opencv/pull/25199/files#r1524623439

ilya-lavrenov commented on modules/dnn/src/op_inf_engine.cpp:
```diff
@@ -10,7 +10,7 @@
 #include <opencv2/dnn/shape_utils.hpp>
 
 #ifdef HAVE_INF_ENGINE
```
```suggestion
#ifdef HAVE_OPENVINO
```

ilya-lavrenov commented on modules/dnn/src/op_inf_engine.hpp:
```diff
@@ -28,6 +28,7 @@
 #define INF_ENGINE_RELEASE_2021_4 2021040000
 #define INF_ENGINE_RELEASE_2022_1 2022010000
 #define INF_ENGINE_RELEASE_2023_0 2023000000
+#define INF_ENGINE_RELEASE_2024_0 2024000000
```
```suggestion
#define OPENVINO_RELEASE_2024_0 2024000000
```
and can we remove old ones?

dkurt replied:
Let’s remove old ones, but keep IE notation to reduce changes.

olpipi replied:
done

dkurt commented on modules/dnn/src/op_inf_engine.hpp:
```diff
@@ -101,7 +90,6 @@ CV__DNN_INLINE_NS_END
-#if INF_ENGINE_VER_MAJOR_GE(INF_ENGINE_RELEASE_2022_1)
 namespace InferenceEngine {
 
 class CNNNetwork {
```
Please remove this workaround with `CNNNetwork` alias

olpipi replied:
done

dkurt commented on modules/dnn/test/test_ie_models.cpp:
```diff
@@ -255,68 +212,90 @@ void runIE(Target target, const std::string& xmlPath, const std::string& binPath
     outputsMap.clear();
-    BlobMap outputBlobs;
-    for (auto& it : net.getOutputsInfo())
+    for (const auto& it : model->outputs())
```
Failed test:
```
[ RUN      ] DNNTestOpenVINO.models/0, where GetParam() = (NGRAPH/CPU, "age-gender-recognition-retail-0013")
Exception message: OpenCV(4.9.0-dev) /home/openvino/opencv/modules/dnn/src/net_impl.cpp:570: error: (-215:Assertion failed) inp.total() in function 'allocateLayers'

/home/openvino/opencv/modules/dnn/test/test_ie_models.cpp:399: Failure
Expected: runCV(backendId, targetId, xmlPath, binPath, inputsMap, cvOutputsMap) doesn't throw an exception.
  Actual: it throws.
runCV
/home/openvino/opencv/modules/dnn/test/test_ie_models.cpp:415: Failure
Expected equality of these values:
  ieOutputsMap.size()
    Which is: 2
  cvOutputsMap.size()
    Which is: 0
unknown file: Failure
C++ exception with description "OpenCV(4.9.0-dev) /home/openvino/opencv/modules/dnn/test/test_ie_models.cpp:419: error: (-215:Assertion failed) dstIt != cvOutputsMap.end() in function 'Body'
" thrown in the test body.
[  FAILED  ] DNNTestOpenVINO.models/0, where GetParam() = (NGRAPH/CPU, "age-gender-recognition-retail-0013") (87 ms)
```

olpipi replied:
fixed

dkurt replied:
well done!

dkurt reviewed (approved):
Great job, many thanks!

olpipi commented:
Hi @dkurt. 
Can we merge this PR or any extra actions are required?

dkurt commented:
@olpipi, Need a comment from @opencv-alalek regarding https://pullrequest.opencv.org/buildbot/builders/4_x_openvino-opencl-skl-lin64/ (uses 2021.4.2 OpenVINO). Is that critical if we merge this PR to 4.x before CI upgrade?

opencv-alalek reviewed (approved):
Well done :+1:

@olpipi Please squash commits into one before merge 

---

OpenVINO builder on legacy CI will be upgraded later.

olpipi commented:
> Well done 👍
> 
> @olpipi Please squash commits into one before merge
> 
> OpenVINO builder on legacy CI will be upgraded later.

Done. Squashed and rebased

CSBVision commented:
After OpenCV's 4.10.0 release including this patch, we tried to upgrade to OpenVINO 2024, too. However, compiling OpenVINO 2024.2.0 with basically only the CPU plugin, linking OpenCV against it and using `model.setPreferableBackend(cv.dnn.DNN_BACKEND_INFERENCE_ENGINE)`, we get the following runtime error in `model.forward()`:

> error: OpenCV(4.10.0) opencv\modules\dnn\src\ie_ngraph.cpp:558: error: (-215:Assertion failed) blobIt != allBlobs.end() in function 'cv::dnn::InfEngineNgraphNet::forward'

Ìt works without any issues if either OpenCV's backend is used or the OpenVINO's 2024 release is replaced by 2023 (we tested 2023.3.0), i.e. it is directly caused by using OpenVINO's 2024 release. We also tested different models, the error is always the same. Any ideas what's causing it?

asmorkalov commented:
@CSBVision could you file a bug with small reproducer and particular OpenVINO versions.
